shrink.sh

```diff
@@ -0,0 +1,152 @@
+#!/bin/bash
+
+FSTAB_FILE='/etc/fstab'
+SHRINK_TAG='x-systemd.shrinkfs'
+
+function read_fstab() {
+    while IFS= read -r line; do
+        case $line in
+        \#* )
+        ;;
+        "" )
+        ;;
+        UUID=*|/dev/* )
+        if [[ "$line" == "$device_name "* ]]; then
+            echo "$line" 
+            break
+        fi
+        ;;
+        esac
+    done < $FSTAB_FILE
+}
+
+
+function get_device_name() {
+    if [[ "$1" == "UUID="* ]]; then
+        parse_uuid "$1"
+    else
+        cut -d " " -f 1 <<< "$1"
+    fi      
+} 
+
+function get_tag_value_in_bytes() {
+    IFS=, read -ra options <<< "$1"
+    for opt in "${options[@]}"; do
+        if [[ "$opt" == "$SHRINK_TAG="* ]]; then
+            /usr/bin/numfmt  --from iec "${opt#*=}" 
+            break
+        fi
+    done
+}
+
+function parse_tag() {
+    if [[ "$1" == *"$SHRINK_TAG="* ]]; then 
+        get_tag_value_in_bytes "$1"
+    fi
+}
+
+function is_device_mounted() {
+    /usr/bin/findmnt --source "$1" 1>&2>/dev/null
+    status=$?
+    if [[  status -eq 0 ]]; then
+        echo "Device $1 is mounted"
+        exit 1
+    fi
+}
+
+function get_current_volume_size() {
+    val=$(/usr/bin/lsblk -b "$1" -o SIZE --noheadings)
+    status=$?
+    if [[ $status -ne 0 ]]; then
+        exit $status
+    fi
+    echo "$val"
+}
+
+function is_lvm(){
+    val=$( /usr/bin/lsblk "$1" --noheadings -o TYPE )
+    status=$?
+    if [[ status -ne 0 ]]; then
+        echo $status
+        exit 1
+    fi
+    if [[ "$val" != "lvm"  ]]; then
+        echo "device $device_name is not of lvm type"
+        exit 1
+    fi
+}
+
+function parse_uuid() {
+    uuid=$(echo "$1"| awk '{print $1}'|awk -F'UUID=' '{print $2}')
+    val=$(/usr/bin/lsblk /dev/disk/by-uuid/"$uuid" -o NAME --noheadings)
+    ret=$?
+    if [[ ret -ne 0 ]]; then
+        echo "Failed to retrieve device name from UUID"
+        exit 1
+    fi
+    echo "/dev/mapper/$val"
+}
+
+
+function shrink_volume() {
+    /usr/sbin/lvreduce --resizefs -L "$2b" "$1"
+}
+
+function process_entry() {
+    expected_size_in_bytes=$(parse_tag "$2")
+    if [[ -z "$expected_size_in_bytes" ]]; then
+        echo "Tag $SHRINK_TAG not found for device '$1' in '$2'"
+        exit 0
+    fi
+    current_size=$(get_current_volume_size "$1")
+    if [[ $current_size -lt $expected_size_in_bytes ]];then
+        echo "Unable to shrink: current volume size of $current_size is lower to expected $expected_size_in_bytes"
```

Comment:
The messages here are misleading. It's not that the script is unable to shrink (which sounds like an error), it's that the volume does not need shrinking because it is already smaller or equal to the requested size

Reply:
Ack. I will remove the prefix `Unable to shrink:`